src/SimpleWMSBody.cpp

```diff
@@ -0,0 +1,552 @@
+  mSun = sun;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+bool SimpleWMSBody::getIntersection(
+    glm::dvec3 const& rayOrigin, glm::dvec3 const& rayDir, glm::dvec3& pos) const {
+
+  glm::dmat4 transform = glm::inverse(getWorldTransform());
+
+  // Transform ray into planet coordinate system.
+  glm::dvec4 origin(rayOrigin, 1.0);
+  origin = transform * origin;
+
+  glm::dvec4 direction(rayDir, 0.0);
+  direction = transform * direction;
+  direction = glm::normalize(direction);
+
+  double b    = glm::dot(origin, direction);
+  double c    = glm::dot(origin, origin) - mRadii[0] * mRadii[0];
+  double fDet = b * b - c;
+
+  if (fDet < 0.0) {
+    return false;
+  }
+
+  fDet = std::sqrt(fDet);
+  pos  = (origin + direction * (-b - fDet));
+
+  return true;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+double SimpleWMSBody::getHeight(glm::dvec2 lngLat) const {
+  // This is why we call them 'SimpleBodies'.
+  return 0;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+glm::dvec3 SimpleWMSBody::getRadii() const {
+  return mRadii;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+bool SimpleWMSBody::Do() {
+  std::lock_guard<std::mutex> guard(mWMSMutex);
+  if (!getIsInExistence() || !pVisible.get()) {
+    return true;
+  }
+
+  cs::utils::FrameTimings::ScopedTimer timer("Simple WMS Bodies");
+
+  if (mActiveWMS.mTime.has_value()) {
+    boost::posix_time::ptime time =
+        cs::utils::convert::toBoostTime(mTimeControl->pSimulationTime.get());
+    for (int i = -mActiveWMS.mPrefetchCount.value_or(0); i <= mActiveWMS.mPrefetchCount.value_or(0);
+         i++) {
+      boost::posix_time::time_duration td = boost::posix_time::seconds(mIntervalDuration);
+      time = cs::utils::convert::toBoostTime(mTimeControl->pSimulationTime.get()) + td * i;
+      boost::posix_time::time_duration timeSinceStart;
+      boost::posix_time::ptime         startTime =
+          time - boost::posix_time::microseconds(time.time_of_day().fractional_seconds());
+      bool inInterval = utils::timeInIntervals(
+          startTime, mTimeIntervals, timeSinceStart, mIntervalDuration, mFormat);
+      if (mIntervalDuration != 0) {
+        startTime -= boost::posix_time::seconds(timeSinceStart.total_seconds() % mIntervalDuration);
+      }
+      std::string timeString = utils::timeToString(mFormat.c_str(), startTime);
+      if (mProperties->mEnableTimespan.get()) {
+        boost::posix_time::time_duration timeDuration =
+            boost::posix_time::seconds(mIntervalDuration);
+        boost::posix_time::ptime intervalAfter = getStartTime(startTime + timeDuration);
+        timeString += "/" + utils::timeToString(mFormat.c_str(), intervalAfter);
+      }
+      auto iteratorText1 = mTextureFilesBuffer.find(timeString);
```

Comment:
Maybe explain/document how things work. 
Thats what i think how it is working:
mTextureFilesBuffer is used to check if file is already in loading state
mTexturesBuffer is used to check if file is already copied to GPU
mTexture holds the real OpenGL textures

i think the names can be better

Reply:
Do you have any recommendations for the names?